DAOS-11218 control: Add --wait flag to dmg pool rebuild commands

mjmac opened the pull request:
Add an opt-in --wait flag to dmg pool exclude/drain/extend/reintegrate.
When supplied, dmg blocks until the rebuild triggered by the operation
reaches the DONE state, polling PoolQuery once per second.

Update a number of ftest sites that used wait_for_rebuild_to_end()
to use the new flag instead, both simplifying the test harness and
ensuring integration test coverage of the new flag.

Signed-off-by: Michael MacDonald <github@macdonald.cx>

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/1/execution/node/1250/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/1/execution/node/1297/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/3/execution/node/703/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/3/execution/node/719/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/4/execution/node/1031/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/4/execution/node/1254/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/4/execution/node/1301/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/5/execution/node/1297/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/5/execution/node/1250/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/6/execution/node/1274/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/6/execution/node/1189/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/7/execution/node/1274/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/7/execution/node/1227/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11600/8/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11600/8/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11600/8/display/redirect

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/9/execution/node/367/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/9/execution/node/468/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/10/execution/node/1033/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/11/execution/node/710/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/11/execution/node/663/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/12/execution/node/1014/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11600/14/testReport/(root)/

daosbuild1 commented:
Test stage Functional on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11600/15/testReport/(root)/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11600/16/testReport/(root)/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/17/execution/node/1283/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/17/execution/node/1256/log

tanabarr commented on src/control/server/mgmt_svc.go:
```diff
@@ -264,6 +289,33 @@ func (svc *mgmtSvc) processBatchedMsgRequests(ctx context.Context, bprChan batch
+
+// serialReqLoop is the main loop for processing serial requests.
+func (svc *mgmtSvc) serialReqLoop(parent context.Context) {
+
```
unnecessary newline

tanabarr reviewed (dismissed):
looks like some good improvements

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/372/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/326/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/357/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.5 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/323/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/299/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/521/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/19/execution/node/1441/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/19/execution/node/1462/log

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11600/20/testReport/

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/20/execution/node/1435/log

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/20/execution/node/1481/log

daosbuild1 commented:
Test stage NLT on EL 8.8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11600/21/testReport/

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/21/execution/node/1491/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11600/21/execution/node/1528/log

tanabarr commented:
I'm assuming this is going to require significant test code changes, is that right?

mjmac commented:
> I'm assuming this is going to require significant test code changes, is that right?

Not sure... I was hoping it would just work, but doesn't seem like it. I did make a small change to the daos_test helper to use the previous async behavior. I will probably need to either also do that in ftest or make changes to handle the synchronous behavior. I'd prefer that most test coverage use the sync path though, as that's the default.

mjmac committed:
DAOS-11218 control: Add --wait flag to dmg pool rebuild commands

Add an opt-in --wait flag to dmg pool exclude/drain/extend/reintegrate.
When supplied, dmg blocks until the rebuild triggered by the operation
reaches the DONE state, polling PoolQuery once per second.

Update a number of ftest sites that used wait_for_rebuild_to_end()
to use the new flag instead, both simplifying the test harness and
ensuring integration test coverage of the new flag.

Signed-off-by: Michael MacDonald <github@macdonald.cx>

daosbuild3 commented:
Test stage Unit Test completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/2/testReport/

daosbuild3 commented:
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/2/testReport/

daosbuild3 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/2/execution/node/1121/log

daosbuild3 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/2/execution/node/1211/log

daosbuild3 commented:
Test stage Functional on SLES 15 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/2/execution/node/1141/log

daosbuild3 commented:
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/4/testReport/

daosbuild3 commented:
Test stage Functional on EL 9 completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/5/execution/node/1064/log

daosbuild3 commented:
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/7/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/9/execution/node/1263/log

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/10/testReport/

daosbuild3 commented:
Test stage NLT completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/11/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/12/testReport/

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/13/testReport/

tanabarr reviewed (dismissed):
go changes LGTM but I think @kccain needs to review to see how it fits in with his recent work.

kccain commented:
> go changes LGTM but I think @kccain needs to review to see how it fits in with his recent work.

Thanks @tanabarr . I will need more time to review this in detail and can follow up with any deeper rebuild comments later (if there are any to add). Initial quick questions and feedback:

- What is the intended exit-code contract for the wait behavior after a successful initial command (reintegrate, drain, extend, exclude)? Should command success be based only on detecting rebuild state done, or should it also consider the final rebuild error value from pool query?
- Should wait stop for both successful and failed rebuild completions, or only for successful completion? I'm most interested in drain, extend, and reintegrate, where failed rebuilds ultimately end in done with a nonzero rebuild error.
- Is there a unit test for the case where rebuild reaches done but reports a nonzero rebuild error?
- What user-facing output is intended while waiting? Nothing until completion, periodic progress/state output, or final-state-only output?
- In ftest coverage for this option, are checks based on machine-readable JSON output or human-readable text parsing? Tom did some work recently to improve the human-readable output to reflect states such as "failing" and "failed" for when a nonzero rebuild error number is seen.

mjmac committed:
Use DerivedState in PoolRebuildResp to determine status

DAOS-18347 added a new DerivedState to the rebuild response
struct. Use this to determine the exit status of a rebuild
rather than the simple State field.

Signed-off-by: Michael MacDonald <github@macdonald.cx>

mjmac commented:
Hi @kccain, thanks for the feedback.

> Initial quick questions and feedback:
> 
> * What is the intended exit-code contract for the wait behavior after a successful initial command (reintegrate, drain, extend, exclude)? Should command success be based only on detecting rebuild state done, or should it also consider the final rebuild error value from pool query?

I've updated the PR to use `DerivedState` as the controlling state. I hadn't seen that change earlier when I revived this patch. So now we'll only get exit 0 on a fully successful operation, not failed or stopped.

> * Should wait stop for both successful and failed rebuild completions, or only for successful completion? I'm most interested in drain, extend, and reintegrate, where failed rebuilds ultimately end in done with a nonzero rebuild error.

I would think it should stop for both, in order to notify the operator if the operation didn't complete successfully for any reason, no?

> * Is there a unit test for the case where rebuild reaches done but reports a nonzero rebuild error?

Yes, added with the rework.

> * What user-facing output is intended while waiting? Nothing until completion, periodic progress/state output, or final-state-only output?

No output until completion. There may be an opportunity to add progress output for long running commands later, but it should be done as part of an overall feature design rather than piecemeal per command. There is a design somewhere for adding tracking support to long running commands that would dovetail well with UX/UI updates.

> * In ftest coverage for this option, are checks based on machine-readable JSON output or human-readable text parsing? Tom did some work recently to improve the human-readable output to reflect states such as "failing" and "failed" for when a nonzero rebuild error number is seen.

The ftest stuff is mostly JSON based in order to avoid fragile text parsing. There should already be unit tests for the pretty printers that produce human-readable output.

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/15/testReport/

mjmac committed:
Merge branch 'master' into mjmac/DAOS-11218

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/16/execution/node/1569/log

mjmac committed:
Merge branch 'master' into mjmac/DAOS-11218

daosbuild3 commented:
Test stage Functional Hardware Large MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/17/execution/node/1796/log

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/17/testReport/

mjmac committed:
Merge branch 'master' into mjmac/DAOS-11218

daosbuild3 commented:
Test stage Functional Hardware Medium Verbs Provider MD on SSD completed with status FAILURE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net//job/daos-stack/job/daos/view/change-requests/job/PR-11600/18/execution/node/1653/log

daltonbohning commented on src/tests/ftest/deployment/network_failure.py:
```diff
@@ -233,11 +233,10 @@ def verify_network_failure(self, ior_namespace, container_namespace):
+        # --wait blocks each reintegrate until its rebuild completes.
         self.log_step("Reintegrate one rank at a time to enable all ranks.")
         for disabled_rank in disabled_ranks:
-            self.pool.reintegrate(ranks=disabled_rank)
-            self.pool.wait_for_rebuild_to_start(interval=5)
-            self.pool.wait_for_rebuild_to_end(interval=10)
+            self.pool.reintegrate(ranks=disabled_rank, wait=True)
```
I think to be more complete we should verify the ranks reintegrated. After the for loop, maybe something like this?
```
output = dmg_cmd.pool_query(pool=self.pool.identifier, show_enabled=True)
disabled_ranks = output["response"].get("disabled_ranks")
if disabled_ranks is not None:
    self.fail(f"Ranks are still disabled after reintegration: {disabled_ranks}")
```

mjmac replied:
Added a query to check for disabled ranks.

daltonbohning commented on src/tests/ftest/deployment/server_rank_failure.py:
```diff
@@ -189,20 +189,15 @@ def verify_rank_failure(self, ior_namespace):
             while True:
                 try:
-                    self.pool.reintegrate(ranks=disabled_rank)
+                    self.pool.reintegrate(ranks=disabled_rank, wait=True)
```
And similar for this test after all reintegration
```
output = self.get_dmg_command().pool_query(pool=self.pool.identifier, show_enabled=True)
disabled_ranks = output["response"].get("disabled_ranks")
if disabled_ranks is not None:
    self.fail(f"Ranks are still disabled after reintegration: {disabled_ranks}")
```

mjmac committed:
Merge branch 'master' into mjmac/DAOS-11218

mjmac committed:
re-query for disabled ranks after reintegration

Signed-off-by: Michael MacDonald <github@macdonald.cx>

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/19/testReport/

mjmac committed:
Merge branch 'master' into mjmac/DAOS-11218

daltonbohning reviewed (commented):
ftest LGTM. Thanks!

daosbuild3 commented:
Test stage Functional Hardware Medium MD on SSD completed with status UNSTABLE.  https://jenkins-3.daos.hpc.amslabs.hpecorp.net/job/daos-stack/job/daos//view/change-requests/job/PR-11600/20/testReport/

mjmac committed:
DAOS-11218 test: fix truthy check

Don't assume that the disabled list will be None; it is more
likely to be an empty list. `if disable_ranks:` works for
either case.

Signed-off-by: Michael MacDonald <github@macdonald.cx>

daltonbohning reviewed (commented):
ftest LGTM

knard38 commented on src/control/cmd/dmg/main.go:
```diff
@@ -91,6 +92,12 @@ func (cmd *cfgCmd) setConfig(cfg *control.Config) {
+// waitCmd is embedded in subcommands that kick off a pool rebuild and may
+// optionally block until that rebuild completes.
+type waitCmd struct {
+	Wait bool `long:"wait" description:"Block until the rebuild triggered by this operation completes"`
```
Would it make sense to replace the boolean `Wait` field with an optional duration instead (e.g. `--wait[=TIMEOUT]`, using go-flags' `optional`/`optional-value` tags), defaulting to blocking indefinitely when no value is supplied so today's behavior is unchanged for existing scripts?

`WaitForPoolRebuild`'s polling loop currently has no upper bound of its own — `cmd.MustLogCtx()` carries no deadline, so a rebuild that never reaches a terminal state would block `dmg` forever — and letting the flag optionally carry a timeout would give users a way to bound that without introducing a second, separate flag.

mjmac replied:
@knard38: This is a good idea. I've pushed an update that incorporates this and your other feedback. Thanks!

knard38 commented on src/control/cmd/dmg/pool.go:
```diff
@@ -567,6 +580,10 @@ func (cmd *poolExcludeCmd) Execute(args []string) error {
 	}
 
+	if err := cmd.maybeWaitForRebuild(resp.HasSuccess()); err != nil {
+		return err
```
From my understanding, the `--json` output selection isn't respected here: returning directly means this never reaches the `if cmd.JSONOutputEnabled() { return cmd.OutputJSON(resp, resp.Errors()) }` branch a few lines below, so a `--wait` failure always falls through to `parseOpts`'s top-level fallback, which reports a `null` response even though `resp` is already populated at this point.
If I am correct, something like this should fix it while still preserving today's plain-text behavior when `--json` isn't set:
```suggestion
		if cmd.JSONOutputEnabled() {
			return cmd.OutputJSON(resp, err)
		}
		return err
```
This applies identically to `poolDrainCmd.Execute` (around line 626) and `poolReintegrateCmd.Execute` (around line 692), which have the same pattern.

knard38 commented on src/control/cmd/dmg/pool_test.go:
If the concern above about `--json` output holds, it'd be good to add a regression test for it here too. From my understanding, none of the current tests can reach that path today.

tanabarr reviewed (approved):
LGTM, happy for this to land once all concerns from Ken and Cedric have been addressed.

kccain commented:
the responses from @mjmac seemed OK from my standpoint
